### PR TITLE
FPGA: Add new sample `component_interfaces_comparison`

### PR DESCRIPTION
**Pull request by KevinUTAT**
# Adding a New Sample(s)

## Description

This sample introduces different invocation/data interfaces that can be used in IP Authoring flow

## Testing

Windows: (https://spetc.intel.com/testanalysis?trview=0&testRunIds=8710728&tapgsize=20)
Linux: https://spetc.intel.com/testanalysis?trview=0&testRunIds=8689034&tapgsize=20

## Checklist
Administrative
- [x] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [x] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [x] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [x] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

**Review by yuguen-intel — APPROVED** [review]

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/streaming-invocation/src/vector_add.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    for (int idx = 0; idx < len; idx++) {
+      int a_val = a_in[idx];
+      int b_val = b_in[idx];
+      int sum = a_val + b_val;
+      c_out[idx] = sum;
+    }
+  }
+};
+
+#define VECT_SIZE 256
```
Comment (pmpeter1) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/streaming-invocation/src/vector_add.cpp:50]: Should avoid usage of #define in C++ and try to use a typesafe alternative like constexpr

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/pipes/src/vector_add.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
+#include <iostream>
+
+// oneAPI headers
+#include <sycl/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include "exception_handler.hpp"
+
+#define VECT_SIZE 256
```
Comment (pmpeter1) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/pipes/src/vector_add.cpp:8]: Should avoid #define in C++ and prefer typesafe alternatives like constexpr

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/naive/src/vector_add.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    for (int idx = 0; idx < len; idx++) {
+      int a_val = a_in[idx];
+      int b_val = b_in[idx];
+      int sum = a_val + b_val;
+      c_out[idx] = sum;
+    }
+  }
+};
+
+#define VECT_SIZE 256
```
Comment (pmpeter1) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/naive/src/vector_add.cpp:27]: Should avoid #define in C++ and use typesafe alternatives

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/csr-pipes/src/vector_add.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class SimpleVAddPipes;
+
+// Forward declare pipe names to reduce name mangling
+class IDPipeA;
+class IDPipeB;
+class IDPipeC;
+
+#define VECT_SIZE 256
```
Comment (pmpeter1) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/csr-pipes/src/vector_add.cpp:17]: Should avoid #define in C++

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/csr-pipes/src/vector_add.cpp (hunk 4 above)
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/csr-pipes/src/vector_add.cpp:17]:
```suggestion
constexpr int kVectorSize = 256;
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/csr-pipes/src/vector_add.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+#elif FPGA_HARDWARE
+    auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+    auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+    // create the device queue
+    sycl::queue q(selector, fpga_tools::exception_handler);
+
+    int count = VECT_SIZE;  // pass array size by value
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/csr-pipes/src/vector_add.cpp:72]:
```suggestion
    int count = kVectorSize;  // pass array size by value
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/mm-host/src/vector_add.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    for (int idx = 0; idx < len; idx++) {
+      int a_val = a_in[idx];
+      int b_val = b_in[idx];
+      int sum = a_val + b_val;
+      c_out[idx] = sum;
+    }
+  }
+};
+
+#define VECT_SIZE 256
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/mm-host/src/vector_add.cpp:58]:
```suggestion
constexpr int kVectorSize = 256;
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/mm-host/src/vector_add.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    // create the device queue
+    sycl::queue q(selector, fpga_tools::exception_handler);
+
+    auto device = q.get_device();
+
+    std::cout << "Running on device: "
+              << device.get_info<sycl::info::device::name>().c_str()
+              << std::endl;
+
+    int count = VECT_SIZE;  // pass array size by value
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/mm-host/src/vector_add.cpp:83]:
```suggestion
    int count = kVectorSize;  // pass array size by value
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/naive/src/vector_add.cpp (hunk 3 above)
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/naive/src/vector_add.cpp:27]:
```suggestion
constexpr int kVectorSize = 256;
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/naive/src/vector_add.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+#elif FPGA_HARDWARE
+    auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+    auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+    // create the device queue
+    sycl::queue q(selector, fpga_tools::exception_handler);
+
+    int count = VECT_SIZE;  // pass array size by value
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/naive/src/vector_add.cpp:47]:
```suggestion
    int count = kVectorSize;  // pass array size by value
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/pipes/src/vector_add.cpp (hunk 2 above)
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/pipes/src/vector_add.cpp:8]:
```suggestion
constexpr int kVectorSize = 256;
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/pipes/src/vector_add.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    // create the device queue
+    sycl::queue q(selector, fpga_tools::exception_handler);
+
+    auto device = q.get_device();
+
+    std::cout << "Running on device: "
+              << device.get_info<sycl::info::device::name>().c_str()
+              << std::endl;
+
+    int count = VECT_SIZE;  // pass array size by value
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/pipes/src/vector_add.cpp:61]:
```suggestion
    int count = kVectorSize;  // pass array size by value
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/streaming-invocation/src/vector_add.cpp (hunk 1 above)
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/streaming-invocation/src/vector_add.cpp:50]:
```suggestion
constexpr int kVectorSize = 256;
```

**Review by ishaghosh27 — APPROVED** [review]

**Pull request merged**